File: src/algorithm/template/layout/field_block/field_block_drawing.py
from src.utils.constants import CLR_BLACK
from src.utils.drawing import DrawingUtils


class FieldBlockDrawing:
    def __init__(self, field_block):
        self.field_block = field_block

    def draw_field_block(self, marked_image, shifted=True, thickness=3, border=3):
        field_block = self.field_block
        # TODO: get this field block using a bounding box of all bubbles instead. (remove shift at field block level)
        FieldBlockDrawing.draw_bounding_rectangle(
            field_block, marked_image, shifted, border
        )

        thickness_factor = 1 / 10
        for field in field_block.fields:
            field.drawing.draw_scan_boxes(
                marked_image, field_block.shifts, thickness_factor, border
            )

        FieldBlockDrawing.draw_field_block_label(
            field_block, marked_image, shifted, thickness
        )

    @staticmethod
    def draw_bounding_rectangle(field_block, marked_image, shifted, border):
        (
            bounding_box_origin,
            bounding_box_dimensions,
        ) = map(
            lambda attr: getattr(field_block, attr),
            [
                "bounding_box_origin",
                "bounding_box_dimensions",
            ],
        )
        block_position = (
            field_block.get_shifted_origin() if shifted else bounding_box_origin
        )
        if not shifted:
            DrawingUtils.draw_box(
                marked_image,
                block_position,
                bounding_box_dimensions,
                color=CLR_BLACK,
                style="BOX_HOLLOW",
                thickness_factor=0,
                border=border,
            )

    @staticmethod
    def draw_field_block_label(field_block, marked_image, shifted, thickness):
        (
            field_block_name,
            bounding_box_origin,
            bounding_box_dimensions,
        ) = map(
            lambda attr: getattr(field_block, attr),
            [
                "name",
                "bounding_box_origin",
                "bounding_box_dimensions",
            ],
        )

        block_position = (
            field_block.get_shifted_origin() if shifted else bounding_box_origin
        )
        text_position = lambda size_x, size_y: (
            int(block_position[0] + bounding_box_dimensions[0] - size_x),
            int(block_position[1] - size_y),
        )
        text = field_block_name
        if shifted:
            text = f"({field_block.shifts}){field_block_name}"

        DrawingUtils.draw_text(marked_image, text, text_position, thickness=thickness)

File: omrchecker-js/packages/core/src/processors/layout/fieldBlock/fieldBlockDrawing.ts
/**
 * Field block drawing class.
 *
 * TypeScript port of src/processors/layout/field_block/field_block_drawing.py
 * Handles visualization of field blocks.
 */

import cv from '../../../utils/opencv';
import { CLR_BLACK } from '../../../utils/constants';
import { DrawingUtils } from '../../../utils/drawing';
import { FieldBlock } from './base';

/**
 * Drawing class for field blocks.
 */
export class FieldBlockDrawing {
  protected fieldBlock: FieldBlock;

  constructor(fieldBlock: FieldBlock) {
    this.fieldBlock = fieldBlock;
  }

  /**
   * Draw field block on marked image.
   *
   * @param markedImage - Image to draw on
   * @param shifted - Whether to use shifted positions
   * @param thickness - Line thickness for text
   * @param border - Border thickness
   */
  drawFieldBlock(
    markedImage: cv.Mat,
    shifted: boolean = true,
    thickness: number = 3,
    border: number = 3
  ): void {
    const fieldBlock = this.fieldBlock;
    // TODO: get this field block using a bounding box of all bubbles instead. (remove shift at field block level)
    FieldBlockDrawing.drawBoundingRectangle(
      fieldBlock,
      markedImage,
      shifted,
      border
    );

    const thicknessFactor = 1 / 10;
    for (const field of fieldBlock.fields) {
      field.drawing.drawScanBoxes(
        markedImage,
        fieldBlock.shifts,
        thicknessFactor,
        border // Pass as number, FieldDrawing handles both Vec3 and number
      );
    }

    FieldBlockDrawing.drawFieldBlockLabel(
      fieldBlock,
      markedImage,
      shifted,
      thickness
    );
  }

  /**
   * Draw bounding rectangle for field block.
   *
   * @param fieldBlock - Field block to draw
   * @param markedImage - Image to draw on
   * @param shifted - Whether to use shifted positions
   * @param border - Border thickness
   */
  static drawBoundingRectangle(
    fieldBlock: FieldBlock,
    markedImage: cv.Mat,
    shifted: boolean,
    border: number
  ): void {
    const boundingBoxOrigin = fieldBlock.boundingBoxOrigin;
    const boundingBoxDimensions = fieldBlock.boundingBoxDimensions;

    if (!boundingBoxOrigin || !boundingBoxDimensions) {
      return;
    }

    const blockPosition = shifted
      ? fieldBlock.getShiftedOrigin()
      : boundingBoxOrigin;

    if (!shifted) {
      DrawingUtils.drawBox(
        markedImage,
        blockPosition,
        boundingBoxDimensions,
        CLR_BLACK,
        'BOX_HOLLOW',
        0,
        border
      );
    }
  }

  /**
   * Draw field block label.
   *
   * @param fieldBlock - Field block to draw
   * @param markedImage - Image to draw on
   * @param shifted - Whether to use shifted positions
   * @param thickness - Text thickness
   */
  static drawFieldBlockLabel(
    fieldBlock: FieldBlock,
    markedImage: cv.Mat,
    shifted: boolean,
    thickness: number
  ): void {
    const fieldBlockName = fieldBlock.name;
    const boundingBoxOrigin = fieldBlock.boundingBoxOrigin;
    const boundingBoxDimensions = fieldBlock.boundingBoxDimensions;

    if (!boundingBoxOrigin || !boundingBoxDimensions) {
      return;
    }

    const blockPosition = shifted
      ? fieldBlock.getShiftedOrigin()
      : boundingBoxOrigin;

    const textPosition = (sizeX: number, sizeY: number): [number, number] => {
      return [
        Math.floor(blockPosition[0] + boundingBoxDimensions[0] - sizeX),
        Math.floor(blockPosition[1] - sizeY),
      ];
    };

    let text = fieldBlockName;
    if (shifted) {
      text = `(${fieldBlock.shifts[0]},${fieldBlock.shifts[1]})${fieldBlockName}`;
    }

    DrawingUtils.drawText(
      markedImage,
      text,
      textPosition,
      undefined, // textSize - use default
      thickness,
      false, // centered
      CLR_BLACK
    );
  }
}

